### ISSUE TITLE
Occasional lockups dumping planet

### ISSUE THREAD

**Issue by tomhughes** [zerebubuth/planet-dump-ng#24]
We've had a couple of instances - one in July (https://github.com/openstreetmap/operations/issues/552) and one this week (https://github.com/openstreetmap/operations/issues/568) where planetdump-ng has experienced some sort of thread deadlock and stopped making progress.

I grabbed a full backtrace of all threads time (https://gist.github.com/tomhughes/250e1504f4689fc31a0ca4e0ab4e029e) and from analysing the output it looks like it had completed nodes and ways and was in the middle of the relations when it stopped.

**Comment by tomhughes** [comment]
We had one good run but it looks like last week locked up again :-(

**Comment by zerebubuth** [comment]
Weird that this is happening so often now. :thinking: I wonder if something changed?

Thanks very much for the thread backtraces, they were very helpful. It looks like one of the output writer threads has died at some point while outputting the relations and this isn't handled properly.

It looks like the [incron script](https://github.com/openstreetmap/chef/blob/0334e3c7306d9798570f1eb476745aa4392d2ab7/cookbooks/planet/templates/default/planetdump.erb) that runs `planet-dump-ng` should send me any output, but the last time it sent me anything was July 2020... So either the output isn't being written to disk, or the script isn't mailing it. If it happens again, please look if there's a `/tmp/planetdump.log.XXX` file. If it's non-empty, then it might give a better clue what's happening.

**Comment by tomhughes** [comment]
I've deployed that, and I think I've figured out the email problem and fixed it. I think you've broken something though because I get a stream of errors now if I try and start the dump:

```
pg_restore: error: one of -d/--dbname and -f/--file must be specified
```

I assume this is something to do with e2d9c704098a85e29d2e0e8a5615617a45a26594?

**Comment by zerebubuth** [comment]
Ooops, fail. Looks like the machine I was testing on had a truly ancient version PostgreSQL (9!). I reverted that commit and pushed a new version, v.1.2.3.

https://github.com/zerebubuth/planet-dump-ng/releases/tag/v1.2.3

**Comment by mnalis** [comment]
So, do the mails with errors arrive now at least?

It seems planet dump has failed again? Today is 2022-09-20, and last one is [planet-220905.osm.pbf](https://planet.openstreetmap.org/pbf/planet-220905.osm.pbf) (created on 2022-09-10).

**Comment by tomhughes** [comment]
Yes it failed, yes I got email, yes our alerts went off, yes we are continuing to investigate.

**Comment by tomhughes** [comment]
As you can see if you look at #25 which is the actual ticket where we're currently dealing with this...

**Comment by tomhughes** [comment]
Which is actually a totally separate issue - it's a crash rather than a lockup.
